### PR TITLE
eeprom: at2x: Add support for at24cXX

### PR DESCRIPTION
**Pull request by kurddt**
Those devices can use several I2C address in order to address more
than 256 Bytes using 8bit addressing. Also several physical component
can be used as a contiguous memory.

Alternative to #21170

Signed-off-by: Guillaume Lager <guillaume.lager@gmail.com>

### CODE REVIEW

File: dts/bindings/mtd/eeprom-spi-i2c.yaml [hunk 1]
```diff
@@ -8,6 +8,7 @@ include: eeprom-base.yaml
 properties:
     size:
       required: true
+      description: EEPROM size in bytes
```
Comment (henrikbrixandersen) [inline dts/bindings/mtd/eeprom-spi-i2c.yaml:11]: This is already present in `dts/bindings/mtd/eeprom-base.yaml`.

File: drivers/eeprom/eeprom_at2x.c [hunk 2]
```diff
@@ -201,29 +209,58 @@ static size_t eeprom_at2x_size(const struct device *dev)
 }
 
 #ifdef CONFIG_EEPROM_AT24
+
+static uint16_t eeprom_at24_translate_offset(const struct device *dev,
+					     off_t *offset)
+{
+	const struct eeprom_at2x_config *config = dev->config;
+
+	const uint16_t addr_incr = *offset >> config->addr_width;
```
Comment (henrikbrixandersen) [inline drivers/eeprom/eeprom_at2x.c:218]: Looks good, thank you. Could you please add a comment (in the source code) explaining that these additions are for handling AT24Cxx as continuous memory?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: dts/bindings/mtd/eeprom-spi-i2c.yaml (hunk 1 above)
Comment (kurddt) [reply dts/bindings/mtd/eeprom-spi-i2c.yaml:11]: Ok I will remove the whole property then

Comment (henrikbrixandersen) [reply dts/bindings/mtd/eeprom-spi-i2c.yaml:11]: No, leave it be. The `size` property is optional in `eeprom-base.yaml` but required in `eeprom-spi-i2c.yaml`.

File: drivers/eeprom/eeprom_at2x.c [hunk 3]
```diff
@@ -201,29 +209,58 @@ static size_t eeprom_at2x_size(const struct device *dev)
...
+}
+
+static size_t eeprom_at24_adjust_read_count(const struct device *dev,
+					    off_t offset, size_t len)
+{
+	const struct eeprom_at2x_config *config = dev->config;
+	const size_t remainder = BIT(config->addr_width) - offset;
+
+	if (len > remainder)
+		len = remainder;
```
Comment (KwonTae-young) [inline drivers/eeprom/eeprom_at2x.c:231]: braces must be added.
> Add braces to every if, else, do, while, for and switch body, even for single-line code blocks.

File: drivers/eeprom/eeprom_at2x.c [hunk 4]
```diff
@@ -404,7 +448,13 @@ static int eeprom_at25_read(const struct device *dev, off_t offset, void *buf,
 		return err;
 	}
 
-	return spi_transceive(data->bus_dev, &data->spi_cfg, &tx, &rx);
+	err = spi_transceive(data->bus_dev, &data->spi_cfg, &tx, &rx);
+
+	if (err < 0) {
```
Comment (KwonTae-young) [inline drivers/eeprom/eeprom_at2x.c:453]: It would be nice if there were no blank lines.

**Review by KwonTae-young — CHANGES_REQUESTED** [review]

File: drivers/eeprom/eeprom_at2x.c (hunk 2 above)
Comment (kurddt) [reply drivers/eeprom/eeprom_at2x.c:225]: I added a comment block before the function

**Comment by kurddt** [comment]
I also made (what I think are) correction to some of the indentation

File: drivers/eeprom/eeprom_at2x.c (hunk 3 above)
Comment (kurddt) [reply drivers/eeprom/eeprom_at2x.c:238]: Check path report the following if I do so

WARNING: braces {} are not necessary for single statement blocks
#71: FILE: drivers/eeprom/eeprom_at2x.c:237:
+	if (len > remainder) {
+		len = remainder;
+	}

File: drivers/eeprom/eeprom_at2x.c [hunk 5]
```diff
@@ -526,7 +583,8 @@ static int eeprom_at2x_init(const struct device *dev)
 		}
 
 		err = gpio_pin_configure(data->wp_gpio_dev, config->wp_gpio_pin,
-					 GPIO_OUTPUT_ACTIVE | config->wp_gpio_flags);
+					 GPIO_OUTPUT_ACTIVE
+						 | config->wp_gpio_flags);
```
Comment (henrikbrixandersen) [inline drivers/eeprom/eeprom_at2x.c:587]: Please leave unrelated changes out.

**Comment by KwonTae-young** [comment]
I have a board with `AT24C04` applied.
```
&i2c1 {
	clock-frequency = <400000>;
	status = "ok";

	eeprom: eeprom@50 {
		compatible = "atmel,at24";
		reg = <0x50>;
		label = "EEPROM";
		size = <512>;
		pagesize = <16>;
		address-width = <9>;
		timeout = <10>;
	};
};
```

When compiling, the following error occurs
```
/work/at24x_test/zephyr/include/toolchain/gcc.h:62:36: error: static assertion failed: "Unsupported address width"
   62 | #define BUILD_ASSERT(EXPR, MSG...) _Static_assert(EXPR, "" MSG)
      |                                    ^~~~~~~~~~~~~~
```

As of now, only `AT24C02` seems to be supported.
Will `AT24C04` support be excluded from this PR?

**Comment by kurddt** [comment]
That's expected, the address width is either 8 or 16.  Setting the size to 512 is enough to handle the  `AT24C04` correctly

**Comment by KwonTae-young** [comment]
@kurddt  I did a simple test.
There seems to be a bit of a problem in the current PR.
I tested it with the `AT24C04`.

**test source**
``` c
/*
 * Copyright (c) 2012-2014 Wind River Systems, Inc.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <zephyr.h>
#include <sys/printk.h>
#include <device.h>
#include <drivers/eeprom.h>

#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
#include <logging/log.h>
LOG_MODULE_REGISTER(test);

void main(void)
{
	const struct device *eeprom;
	size_t eeprom_size;
	const uint8_t wr_buf[17] = {0x00, 0x01, 0x2, 0x03, 0x04, 0x05, 0x06, 0x07,
				0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F,
				0x10};
	uint8_t rd_buf[17] = {0x00, };
	int ret;

	/* EEPROM */
	eeprom = device_get_binding(DT_LABEL(DT_ALIAS(eeprom_0)));
	if (!eeprom) {
		LOG_ERR("Cannot get EEPROM device");
		return;
	}

	eeprom_size = eeprom_get_size(eeprom);
	LOG_INF("eeprom_size: %d", eeprom_size);

	/* 8-biyte read/write */
	ret = eeprom_write(eeprom, 0, wr_buf, 8);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}

	memset(rd_buf, 0, sizeof(rd_buf));
	ret = eeprom_read(eeprom, 0, rd_buf, 8);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}
	
	if (memcmp(wr_buf, rd_buf, 8)) {
		LOG_ERR("8-byte write, read mismatch.");
		LOG_HEXDUMP_ERR(wr_buf, 8, "wr_buf");
		LOG_HEXDUMP_ERR(rd_buf, 8, "rd_buf");
	}

	/* 12-biyte read/write */
	ret = eeprom_write(eeprom, 0, wr_buf, 12);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}

	memset(rd_buf, 0, sizeof(rd_buf));
	ret = eeprom_read(eeprom, 0, rd_buf, 12);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}
	
	if (memcmp(wr_buf, rd_buf, 12)) {
		LOG_ERR("12-byte write, read mismatch.");
		LOG_HEXDUMP_ERR(wr_buf, 12, "wr_buf");
		LOG_HEXDUMP_ERR(rd_buf, 12, "rd_buf");
	}


	/* 16-biyte read/write */
	ret = eeprom_write(eeprom, 0, wr_buf, 16);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}

	memset(rd_buf, 0, sizeof(rd_buf));
	ret = eeprom_read(eeprom, 0, rd_buf, 16);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}
	
	if (memcmp(wr_buf, rd_buf, 16)) {
		LOG_ERR("16-byte write, read mismatch.");
		LOG_HEXDUMP_ERR(wr_buf, 16, "wr_buf");
		LOG_HEXDUMP_ERR(rd_buf, 16, "rd_buf");
	}

	/* 17-biyte read/write */
	ret = eeprom_write(eeprom, 0, wr_buf, 17);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}

	memset(rd_buf, 0, sizeof(rd_buf));
	ret = eeprom_read(eeprom, 0, rd_buf, 17);
	if (ret) {
		LOG_ERR("eeprom_write() failed.");
		return;
	}
	
	if (memcmp(wr_buf, rd_buf, 17)) {
		LOG_ERR("17-byte write, read mismatch.");
		LOG_HEXDUMP_ERR(wr_buf, 17, "wr_buf");
		LOG_HEXDUMP_ERR(rd_buf, 17, "rd_buf");
	}

	LOG_INF("EEPROM Test End!");

	return;
}
```

## 1. Current PR(address-width = <16>)
```
&i2c1 {
	clock-frequency = <400000>;
	status = "ok";

	eeprom: eeprom@50 {
		compatible = "atmel,at24";
		reg = <0x50>;
		label = "EEPROM";
		size = <512>;
		pagesize = <16>;
		address-width = <16>;
		timeout = <10>;
	};
};
```

**console**
```
[00:00:00.005,000] <inf> test: eeprom_size: 512                                                     
[00:00:00.017,000] <err> test: 16-byte write, read mismatch.                                        
[00:00:00.017,000] <err> test: wr_buf                                                               
00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 0f |........ ........                                 
[00:00:00.017,000] <err> test: rd_buf                                                               
00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 10 |........ ........                                 
[00:00:00.025,000] <err> test: 17-byte write, read mismatch.                                        
[00:00:00.025,000] <err> test: wr_buf                                                               
00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 0f |........ ........                                 
10                                               |.                                                 
[00:00:00.025,000] <err> test: rd_buf                                                               
10 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 10 |........ ........                                 
10                                               |.                                                 
[00:00:00.025,000] <inf> test: EEPROM Test End!
```





## 2. Other PR(address-width = <9>)
PR: #21170 
```
&i2c1 {
	clock-frequency = <400000>;
	status = "ok";

	eeprom: eeprom@50 {
		compatible = "atmel,at24";
		reg = <0x50>;
		label = "EEPROM";
		size = <512>;
		pagesize = <16>;
		address-width = <9>;
		timeout = <10>;
	};
};
```

**console**
```
[00:00:00.003,000] <inf> test: eeprom_size: 512                                                     
[00:00:00.023,000] <inf> test: EEPROM Test End!
```

**Comment by kurddt** [comment]
@KwonTae-young The address-width is meant to be used as the word address width, for `AT24C04` it's 8 bit. It needs to be set to 16 for chips like `AT24CM01`.

Basically 8 means 1 byte address will be sent on the I2C bus after the device addressing. When set to 16, 2 bytes of address are sent

**Comment by KwonTae-young** [comment]
@kurddt I confirmed that it works with the address-width set to 8.
```
/ {
	aliases {
		eeprom-0 = &eeprom;
	};
};

&i2c1 {
	clock-frequency = <400000>;
	status = "ok";

	eeprom: eeprom@50 {
		compatible = "atmel,at24";
		reg = <0x50>;
		label = "EEPROM";
		size = <512>;
		pagesize = <16>;
		address-width = <8>;
		timeout = <10>;
	};
};
```

I thought the address-width should be 9-bit as shown in the datasheet.
image

Currently, AT24C04 is controlling the upper address by controlling `bus_addr`.
Some people think it can be confusing.
First of all it works well. :smile:

**Review by KwonTae-young — APPROVED** [review]

**Comment by kurddt** [comment]
Yes I looked at the Linux kernel to get inspiration on how to deal with that

**Review by henrikbrixandersen — APPROVED** [review]
Looks good. Thanks!

**Comment by kurddt** [comment]
Might be good to remove the "Device Tree" and "Device Tree bindings" tags since they are not relevant anymore. I seems that I can't do that myself

**Review by carlescufi — APPROVED** [review]
